akka-stream/src/main/scala/akka/stream/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,212 @@
+ *
+ * We do not materialize the refs back and forth, which is why the 2nd param.
+ */
+final class SinkRef[In] private[akka] (
+  private[akka] val initialPartnerRef:       OptionVal[ActorRef],
+  private[akka] val canMaterializeSourceRef: Boolean
+) extends akka.stream.javadsl.SinkRef[In] {
+  import akka.stream.StreamRefs._
+
+  val in = {
+    val inletName = s"${Logging.simpleName(getClass)}($initialRefName).in"
+    Inlet[In](inletName)
+  }
+  override def shape: SinkShape[In] = SinkShape.of(in)
+
+  private def initialRefName =
+    if (initialPartnerRef.isDefined) initialPartnerRef.get
+    else "<no-initial-ref>"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val promise = Promise[SourceRef[In]]
+    if (!canMaterializeSourceRef) promise.failure(StreamRefs.CyclicMaterializationAttemptException(
+      "SinkRef", otherStage = "SourceRef"))
+
+    val logic = new TimerGraphStageLogic(shape) with StageLogging with InHandler {
+
+      private[this] lazy val streamRefsMaster = StreamRefsMaster(ActorMaterializerHelper.downcast(materializer).system)
+
+      // settings ---
+      import StreamRefAttributes._
+      private[this] lazy val settings = streamRefsMaster.settings
+
+      private[this] lazy val subscriptionTimeout = inheritedAttributes
+        .get[StreamRefAttributes.SubscriptionTimeout](SubscriptionTimeout(settings.subscriptionTimeout))
+      // end of settings ---
+
+      override protected lazy val stageActorName: String = streamRefsMaster.nextSinkRefName()
+      private[this] var self: GraphStageLogic.StageActor = _
+      implicit def selfSender: ActorRef = self.ref
+
+      private var partnerRef: OptionVal[ActorRef] = OptionVal.None
+      private def getPartnerRef: ActorRef =
+        if (partnerRef.isDefined) partnerRef.get
+        else throw StreamRefs.TargetRefNotInitializedYetException()
+
+      val SubscriptionTimeoutTimerKey = "SubscriptionTimeoutKey"
+
+      // demand management ---
+      private var remoteCumulativeDemandReceived: Long = 0L
+      private var remoteCumulativeDemandConsumed: Long = 0L
+      // end of demand management ---
+
+      private var completedBeforeRemoteConnected: OptionVal[Try[Done]] = OptionVal.None
+
+      override def preStart(): Unit = {
+        self = getStageActor(initialReceive)
+        if (initialPartnerRef.isDefined) // this will set the `partnerRef`
+          observeAndValidateSender(initialPartnerRef.get, "Illegal initialPartnerRef! This would be a bug in the SinkRef usage or impl.")
+
+        log.debug("Created SinkRef, pointing to remote Sink receiver: {}, local worker: {}", initialPartnerRef, self.ref)
+
+        if (canMaterializeSourceRef)
+          promise.success(new SourceRef(OptionVal(self.ref), canMaterializeSinkRef = false))
+
+        if (partnerRef.isDefined) {
+          getPartnerRef ! StreamRefs.OnSubscribeHandshake(self.ref)
```

Comment:
So, this message is only sent, if the other side was created first? I guess it makes sense because otherwise the other side can send demand right away.